README.md
<p align="center">
 <img width="100px" src="https://media.discordapp.net/attachments/941663439879802893/947385760183627786/YZB-5.png" align="center" alt="Yuza" />
 
 <h2 align="center">[=|  Yuza  |=]</h2>

 <p align="center" class="desp"> 一個簡單、有趣、實用的Discord機器人 </p>

 <p align= "center">
 <a href="https://github.com/Youzi9601">
      <img alt="作者：Youzi" src="https://img.shields.io/badge/%E4%BD%9C%E8%80%85-Youzi-yellow?style=for-the-badge&logo=Github" />
    </a>
     <p align="center"> 
     </p>

>   <p>架設&使用環境
>   </p>
>
>   <a href="https://nodejs.org/">
>    <img alt="Node.js" src="https://img.shields.io/badge/Node.js-v16.*~v19.*-blue.svg?style=for-the-badge&logo=node.js" /></a>
>
>   <a href="https://www.npmjs.com/">
>    <img alt="NPM" src="https://img.shields.io/badge/NPM-v16.0.0+-blue.svg?style=for-the-badge&logo=npm" /></a>
>
> <img alt="語言" src="https://img.shields.io/github/languages/top/Youzi9601/YZBot?logo=javascript&style=for-the-badge" /></a>

<p align="center"> 
     </p>

>   <p>資訊
>   </p>
>
> [![資訊卡](https://github-readme-stats.vercel.app/api/pin/?username=Youzi9601&repo=YZBot&locale=cn&theme=apprentice)](https://github.com/Youzi9601/YZBot/)

<p align="center"> 
     </p>
     
> 
>   <p>聯絡我們
>   </p>
> 
>  <a href="https://discord.gg/Vq3F8DUNzf">
>    <img src="https://img.shields.io/discord/849809683085525032?style=for-the-badge&logo=discord&label=%E6%94%AF%E6%8F%B4%E4%BC%BA%E6%9C%8D%E5%99%A8" alt="前往Discord"></a>
>
<details>
<summary>其他資訊</summary>
<a href="https://github.com/Youzi9601/YZBot/blob/master/package.json">
<img src="https://img.shields.io/github/package-json/v/Youzi9601/YZBot?label=%E7%89%88%E6%9C%AC&logo=github&style=for-the-badge"></a>
<a href="https://github.com/Youzi9601/YZBot/issues">
<img src="https://img.shields.io/github/issues/Youzi9601/YZBot?style=for-the-badge&logo=github&label=%E5%95%8F%E9%A1%8C" alt="GitHub 問題回報" /></a>
<a href="https://github.com/Youzi9601/YZBot/pulls"><img src="https://img.shields.io/github/forks/Youzi9601/YZBot?style=for-the-badge&logo=github&label=%E5%88%86%E6%94%AF" alt="GitHub 分叉"  /></a>
<img src="https://img.shields.io/github/downloads/Youzi9601/YZBot/total?label=%E4%B8%8B%E8%BC%89%E6%95%B8%E9%87%8F&logo=github&style=for-the-badge" alt="下載數量"></a> 
<img src="https://img.shields.io/github/watchers/Youzi9601/YZBot?label=%E8%B7%9F%E8%B9%A4&logo=Github&style=for-the-badge" alt="追蹤數"></a> 
<img src="https://img.shields.io/github/license/Youzi9601/YZBot?label=%E8%A8%B1%E5%8F%AF%E8%AD%89&logo=Github&style=for-the-badge" alt="許可證"></a> 
<img src="https://img.shields.io/github/languages/code-size/Youzi9601/YZBot?label=%E4%BB%A3%E7%A2%BC%E5%A4%A7%E5%B0%8F&logo=github&style=for-the-badge" alt="代碼大小"></a> 
<img src="https://img.shields.io/github/stars/Youzi9601/YZBot?label=%E8%A9%95%E5%88%86&logo=Github&style=for-the-badge" alt="評分"></a> 
<img src="https://img.shields.io/circleci/build/github/Youzi9601/YZBot?label=%E6%9C%8D%E5%8B%99%E6%B8%AC%E8%A9%A6&logo=circleci&style=for-the-badge&token=2998169653316f89caa76cda45c0c8fff9a48983"></a></a>
</details>

---

<p align="center">

</p>

# 目錄

### 注意：目前機器人重新撰寫中，所以無法使用！

-   [介紹](#介紹)
-   [安裝並啟動](#安裝並啟動)
-   [貢獻你的想法](#貢獻你的想法)

# 介紹

一個由 Youzi9601 所製作的機器人，
內有各種管理命令以及其他有趣的小功能。

# 安裝並啟動

<details>

  <summary>安裝方式(點我打開)</summary>

1. 下載最新的檔案 [`(點我下載)`](https://github.com/Youzi9601/YZBot/archive/refs/heads/master.zip) ，並解壓縮

2. 解壓縮，並於**終端機**執行 `npm install`

3. 將 `Config.example.js` 改成 `Config.js`

4. 將您的機器人設定填入 `Config.js` & `.env` 中 (內有使用說明)
   ( token 以及 Client ID 可以從 <https://discord.com/developers/applications> 找到 )

5. 運行 `Bot.js` (執行 `node bot.js` 或 `node .`) 讓機器人註冊指令並開始運行，此時機器人會上線 (注意：註冊命令需要一些時間！)。

</details>

# 製作日誌

<details>

  <summary>機器人的更新日誌(點我打開)</summary>

### 正式發布

> v1.0.0 - 尚未發布

### 測試版本

> v0.0.46 - [Change Log](https://github.com/Youzi9601/YZBot/blob/master/CHANGELOG.md)

</details>

# 貢獻你的想法

<details>

  <summary>貢獻你的想法吧！(點我打開)</summary>
如果想要把你的想法貢獻給我們，請先注意：

1. 拉取檔案後，如果沒有安裝 `npm-check-updates` 或 `eslint` 套件，請在終端機執行 `npm run dev` 來安裝套件

2. 在終端機執行指令: `npm run check` ，如果 ESLint 報出錯誤，則依 ESLint 的提示修改。修改完後執行 `npm run lint` ，重複此步驟直到 ESLint 沒有報錯為止。

3. 恭喜！你可以到 [這裡](https://github.com/Youzi9601/YZBot/pulls) 開啟一個 Pull Request，讓我們知道你的想法！

</details>

# 感謝

-   使用這個東西的人
-   我自己的努力
-   幫助我的人


CHANGELOG.md
# 版本更改紀錄

所有的相關改動將會放在這邊。

---

# 正式版

## v1.0.3

-   新增音樂系統

---

# 測試版

## v1.1.0

-   新增 井字棋 對戰&AI 系統

## v0.0.50

即將發布

## v0.0.49 - 當前版本

### 漏洞修復

-   修正 Web 問題 ([fd2bb47](https://github.com/Youzi9601/YZBot/commit/fd2bb47dd7a3bd66d7d23c282b27c2aa43066789))

### 新功能

-   資料庫系統整合(mongoose & 本地儲存 & json.sqlite)
-   插件系統新增&整合
-   `機器人命令#fun系列` 新增
-   `機器人事件` 分配調整
-   `機器人命令` 本地化
-   `Web` 網站架設
-   `Readme.md` 檔案撰寫

---


Config.js
require('dotenv').config();
module.exports = {
    /**
     *
     * Config 設定檔案
     * > 這是一個設定檔案，裡面的相關設定可以自由填寫！
     *
     * 格式：
     * // 敘述內容
     * 設定名稱: process.env.設定名稱(我們預設的) || "您要填寫的內容"
     * (process.env.設定名稱 的設定檔案位在 ".env" 檔案裡面！)
     * 部分設定會希望放在".env"中，以免重要的密鑰等訊息遭到濫用！
     *
     * 建議：
     * 可多加利用 伺服器->設定->整合->機器人整合
     * 來調整相關命令的使用權限喔！
     *
     */


    /**
     * 基本設定
     */
    // 檔案&安裝檔 更新
    update: {
        // 自動更新
        auto: process.env.update_auto || false,
        // 自動安裝所有的npm包?
        install_package: process.env.update_install_package || true,
    },

    // 機器人設定
    bot: {
        // 啟動時會使用這個金鑰登入 (也就是機器人token)
        token: process.env.bot_token,

        // 機器人註冊 [ / ] 斜線命令 會使用這個機器人ID (填入機器人ID 或者Developer Portal的Application ID，其實都一樣。)
        clientID: process.env.bot_clientID,
        clientSECRET: process.env.bot_clientSECRET,

        // 機器人的權限(用數字表示)
        permissionID: Number(process.env.bot_clientPermissionID) || 543312702935,

        // 機器人狀態
        botPresence: {
            activities: [
            /**
             * 例：
             *
             * {
             * // 名稱
             *   name: `狀態名稱`,
             * // 類別，可為："PLAYING","STREAMING","LISTENING","WATCHING","CUSTOM","COMPETING"
             *   type: `類別`,
             * // 如果type為"Streaming"可使用以下來輸入網址！(如果沒有則無效)
             *   url: `網址`,
             * }
             *
             * // 可用變數：
             * {bot.name} - 機器人名字
             * {count.guilds} - 伺服器數量
             * {count.members} - 使用者數量
             * {Youtube.subs} - Youtube 訂閱者數量 (需要在 "Youtube.url" 的選項中輸入連結！)
             */

                {
                    'type': 'WATCHING',
                    'name': '{count.guilds}個伺服器&{count.members}個使用者',
                },
                {
                    'name': '訂閱 Youzi9601 ！',
                    'type': 'STREAMING',
                    'url': 'https://www.twitch.tv/Youzi9601',
                },
                {
                    'name': 'Youzi9601 訂閱數：{Youtube.subs}位！',
                    'type': 'COMPETING',
                },

            ],
            // 狀態
            status: 'idle',
        },
    },

    // 機器人開發者們的ID，有允許使用 開發者 指令，並且可以跳過大多伺服器的命令執行限制(如冷卻)
    // 例: ["856918496893599805","862347263690539009","849809683085525032"]
    developers:
        ((`${JSON.parse(process.env.developers || "[]")}` != "") ?
            JSON.parse(process.env.developers || "[]") :
            undefined) ||
        [
            '856918496893599805',
        ],

    // 機器人訊息命令前綴(訊息觸發，允許多個前綴！)
    // 例: ["前綴1","前綴2"]
    // 使用方式: "!ping"、">ping"
    prefix:
        ((`${JSON.parse(process.env.prefix || "[]")}` != "") ?
            JSON.parse(process.env.prefix || "[]") :
            undefined) ||
        [
            '>',
            '/',
            '!>',
        ],


    /**
     *
     * 分片系統(使用同一個機器人但為伺服器分區，減少機器人過多的執行問題)
     *
     * *注意：這必須使用！
     * 用於：2000個伺服器以上的機器人
     *
     */
    sharding: {
        // 生成數量(預設自動)
        amount: process.env.sharding_amount || 'auto',
        // 是否在控制台中輸出 分片廣播 所回傳的全部分片資料(不啟用，因為紀錄會太多)
        logFetchClientValues: false,
    },

    /**
     * 網頁
     */
    // 網頁
    web: {
        noWeb: process.env.web_noWeb || true,
        // 網站位置
        domain: process.env.web_domain || 'http://localhost',
        port: process.env.web_port || 80,
        // 是否顯示Port？（於網址中）
        show_port: process.env.web_show_port || true,

        // 網頁的加密&解密 (secret)
        secret: process.env.web_secret || 'secret is cat (owo!',
        // 網站的連結
        links: {
            github: 'https://github.com/Youzi9601',
            discord: 'https://discord.gg/HKekqu9hWW',
        },
        // 網站最大上限數量於時間內
        maxVisitor: {
            count: process.env.web_maxVisitorCount || 3000,
            minute: process.env.web_maxVisitorInMinute || 10,
        },
        // 網頁接收&傳輸設定
        webhook: {
            // 投票接收系統
            vote: {
                /**
                 * 投票控制選項&驗證
                 */
                // Dst 的投票驗證
                dst: {
                    authorization: process.env.webhook_vote_dst_authorization || 'test_vote',
                },
                // 敬請期待更多投票接收系統的整合

                /**
                 * 傳回的頻道
                 */
                channel: process.env.webhook_channel || '976709522016043058',
            },
        },
    },


    // 資料庫
    database: {
        // 你所使用的儲存方式，有"mongoose","quick.db"(json.sqlite),"locateStorge"(本地json檔案),"none"(不使用)這幾個選項
        // 但是，Quick.db還是會用於小部分檔案儲存，這些小檔案是不影響機器人運做的(例如：分片系統狀態警報)
        use: process.env.database_use || "locateStorge",
        // 請放置網址於下方，用來連接到資料庫
        MongoDB: process.env.database_MongoDB || "",
    },


    /**
     *
     * 日誌輸出&日誌控制設定
     *
     */
    // 控制台日誌文件紀錄天數(單位：天，等同 ./logs/ 的子目錄下的檔案數量｜-1 為永久保存紀錄｜0 為不保存紀錄｜過了這時間以後就會自動刪除)
    logs_clear: process.env.logs_clear || 3,
    console: {
        // 是否報告 Discord Error 錯誤訊息
        error: process.env.console_console_error || true,
        // 是否報告 Discord Warn 警告訊息
        warn: process.env.console_console_warn || true,
        // 是否報告 Discord debug 除錯訊息
        debug: process.env.console_console_debug || false,
    },
    /**
     * 機器人基本傳輸頻道設定
     */
    guild: {
        // 伺服器邀請代碼 (https://discord.gg/${邀請代碼})
        invite_code: process.env.guild_invite_code || 'Vq3F8DUNzf',
        // 紀錄伺服器的ID
        ServerID: process.env.guild_serverid || '769587106300428328',
        // 紀錄伺服器的相關頻道
        Channels: {
        // 機器人所有記錄會在這個頻道 (填入ID)
            All: process.env.guild_Channels_All || '940610972454912042',
            // 機器人啟動時會記錄在這個頻道 (填入ID)
            ClientStatus: process.env.guild_Channels_ClientStatus || '940610972454912042',
            // 機器人啟動時會記錄在這個頻道 (填入ID)
            ClientOn: process.env.guild_Channels_ClientOnChannel || '940610972454912042',
            // 若有錯誤的回報(各種)會記錄在這個頻道 (填入ID)
            report: process.env.guild_Channels_reportChannel || '940610972454912042',
            // 機器人指令的使用紀錄會記錄在這個頻道 (填入ID)
            commandRec: process.env.guild_Channels_commandRecChannel || '940610972454912042',
            // 該伺服器的變動被記錄在這個頻道 (填入ID)
            serverRec: process.env.guild_Channels_serverRecChannel || '940610972454912042',
            // 機器人被邀請進入伺服器的紀錄會記錄在這個頻道 (填入ID)
            inviteChannel: process.env.guild_Channels_inviteChannel || '940610972454912042',
        },
    },

    /**
     * 雜項
     */

    // 調整時差
    // 在GMT-8區為"-8",在GMT+8為"8"
    GMT: process.env.GMT || 8,
    // Youtube 頻道，用於將個人訂閱人數顯示於機器人的狀態中(目前暫時不會用)
    youtube: 'https://www.youtube.com/channel/UCGLbazmDlVg22pO6aGuSRsw',

    /**
     * 插件(擴充功能)系統
     * *這邊的 client 控制選項不一定會在 Root/plugins/ 中，有些是整合至機器人事件/命令接收裡面！(會被歸類為 client )
     * *shared 分片系統所執行的程序通常為定期讀取API或是一些每個伺服器都會用的定時播報系統。
     */
    plugins: {
        client:{
            // 抽獎系統
            giveaways: {
                // 是否要顯示創辦者?
                host_user: true,
                // 提及所有人?
                everyoneMention: false,
            },

            // Openai ChatGPT 的功能，可讓機器人新增以下命令：
            // "/chatgpt question:你的問題"
            openai: {
                use: process.env.openai_use || false,
                openaiKEY: process.env.openai_openaiKEY || '',
            },
        },
        shared: {
            // 地震 API
            Earthquake: {
                // 請從中央氣象局資料庫開放平台中取得一個 API KEY
                TaiwanCWBAPIkey: process.env.Earthquake_TaiwanCWBAPIkey || '',
            },
        },
    },


    /**
     *
     * 開發(測試)版本
     * // 這功能絕大部分都不能使用！
     */
    beta: {
        // RPC為Discord遊戲狀態顯示系統，目前為測試版
        nothing: '沒東西',
    },
};
